### DISCUSSION TITLE
Structured MDC: avoid stale values, and make the code easier to read

### DISCUSSION THREAD

**Discussion by vlsi** [qos-ch/slf4j#297] [Ideas]
The current MDC has no clear lifecycle boundaries.
In other words, if somebody modifies MDC, then all the subsequent code inherits the values.

The current MDC is relatively easy to implement (just a HashMap would work), however
* current MDC often results in "stale values hanging in MDC". I often see a pattern where users add their own workarounds like try-finally that removes values from MDC.
* current MDC is incompatible with Kotlin's coroutines which embrace structured concurrency, thus they explicitly forbid "unexpected inherited values outside of the scope"

It would be nice if MDC in Slf4j was structured. In other words, code structure should provide a clear vision on the visibility scope of the values in MDC.

For instance, here's a sample of the code with the current slf4j:

```java
void handleHttpRequest(HttpRequest req) {
    MDC.put("user", req.getSession().get("username"));
    if (req.getPath().equals("/status")) {
        handleStatusRequest(req);
    } else if (...) {...}

    // Note: at this point, MDC does contain "requestId" which is surprising
}

void handleStatusRequest(HttpRequest req) {
    MDC.put("requestId", req.getSession().get("username"))
    ... 
}
```

If would be nice if it were forbidden to "update MDC values via MDC.put".

For example:

```java
try (var mdcScope = MDC.add("requestId", requestId)) {
   // code inside try would see requestId
}
// requestId won't be visible here
```

**Comment by ceki** [comment]
Are you aware of [MDC.putClosable](https://www.slf4j.org/apidocs/org/slf4j/MDC.html#putCloseable-java.lang.String-java.lang.String-) ?

**Reply by prdoyle** [reply]
Could `put` just be an alias for `putCloseable`?
